### DISCUSSION TITLE
Docker build environment for CI/CD pipeline

### DISCUSSION THREAD

**Discussion by mattheys** [advplyr/audiobookshelf-app#1349] [General]
Has anyone had any success creating a Docker build environment to output built APK's.  I'm looking at building the Android app for myself from the latest source in a consistent way without having to install everything locally.

At the moment it's failing because I need to provide signing keys, I don't know if that's my last issue or just the next one to get past.  So if anyone has already done this I'd be grateful.

```yaml
FROM node:20

WORKDIR /build
RUN git clone https://github.com/advplyr/audiobookshelf-app .
RUN npm install
RUN npm run generate
RUN npx cap sync

RUN apt update && apt install default-jre android-sdk sdkmanager -y
ENV ANDROID_HOME=/usr/lib/android-sdk

RUN yes | sdkmanager --licenses
RUN npx cap build android --androidreleasetype APK
```

**Comment by advplyr** [comment]
I haven't but we have a github workflow that builds APKs you can look at.
https://github.com/advplyr/audiobookshelf-app/blob/master/.github/workflows/build-apk.yml

**Comment by mattheys** [comment]
Sorry for opening it as an issue, and thank you, I'll take a look at the workflow.  I'll have to learn GitHub workflow, we use Azure Pipelines at work.

**Comment by mattheys** [comment]
Thank you, it built :-) I haven't loaded it on my phone yet as I need to work out how to get it off the LXC and onto my phone! I'll do some work on polishing it up a bit and share it for anyone else who might want to fork or run more recent builds and then share back.

**Comment by mattheys** [comment]
So it's going well,  You can find it at https://github.com/mattheys/abs-builder 

I'd like to push the gradle install into the docker image to speed up multiple builds but can't tell from the gradlew file what is actually calling the install! 

I think I can get around that by creating a named container and just repeatedly starting it to keep some state.  Kind of goes against Docker principals but it's a work around for now and nothing stops the image from working new each time, it's just a little slower.

The reason I'm doing this is to try and test out the custom headers on the connection screen without installing lots of software on my desktop.  I guess forking the repo and running the github workflow would have probably done the same.  However if anyone else wants an easy-ish way of building the Android App then it might help them.

There are Environment variables to set the Repo and Branch if you are testing out your own forks, be sure to set these in the docker create if you are doing it that way so they are present when doing docker start.

I don't think this is the most elegant way of doing it, however Java, Gradle, Vue, Nuxt and Capacitor aren't something I'm particularly familiar with.  If anyone has any suggestions, please open an issue or pull request.

`docker create --name builder -e ACCEPT_LICENSE=Y -v ./:/output mheys1/abs-builder:latest`
`docker start builder && docker logs -f builder`

**Comment by nrjohnstone** [comment]
This looks handy, exactly what I was looking for.. It would be great if a test apk was built and provided for every merge, but at least this means I can do it locally myself... Just wondering if you are still using this ?
